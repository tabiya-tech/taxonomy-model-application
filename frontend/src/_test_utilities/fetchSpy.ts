export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
