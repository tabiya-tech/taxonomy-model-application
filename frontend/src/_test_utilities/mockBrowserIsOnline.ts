import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
