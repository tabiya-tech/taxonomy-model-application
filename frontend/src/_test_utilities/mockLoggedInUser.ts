import { AuthContextValue, TabiyaUser, TabiyaUserRole, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>){
  let user: TabiyaUser = mockedValues.user || { username: "foo", roles: [] };
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: TabiyaUserRole) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues
  });
}
