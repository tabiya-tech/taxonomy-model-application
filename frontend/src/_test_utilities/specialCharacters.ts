const sampleValidSpecialCharacters = "^°!\"§$%&/()=?`´*+#'-_.,;<>öäü@[]{}|¡“¶¢[]≠¿'„…∞~<•±æœ";
const sampleInternationalCharacters = "αβγδεζηθικλμνξοπρστυφχψω";
const THREE_BYTE_UTF8_CHAR = "€";
export const WHITESPACE = " \n\r\t";

export function getTestString(length: number, prefix: string = ""): string {
  return (prefix + sampleValidSpecialCharacters + sampleInternationalCharacters)
    .slice(0, length)
    .padEnd(length, THREE_BYTE_UTF8_CHAR);
}

export function getRandomString(length: number) {
  let result = "";
  const characters = sampleValidSpecialCharacters + sampleInternationalCharacters;
  const charactersLength = characters.length;
  let counter = 0;
  while (counter < length) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength)); // NOSONAR
    counter += 1;
  }
  return result;
}

export function getRandomLorem(length: number) {
  let result = "";
  const characters = "     abcdefghijklmnopqrstuvwxyz"; // 5 spaces for 20% chance of whitespace in the english language
  const charactersLength = characters.length;
  let counter = 0;
  while (counter < length) {
    const newChar = characters.charAt(Math.floor(Math.random() * charactersLength)); // NOSONAR
    if (!(newChar === " " && result.charAt(result.length))) {
      result += newChar;
      counter += 1;
    }
  }
  return result;
}
