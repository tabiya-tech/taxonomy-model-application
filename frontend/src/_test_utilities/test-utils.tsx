// Based on https://testing-library.com/docs/react-testing-library/setup/

import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/providers";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
