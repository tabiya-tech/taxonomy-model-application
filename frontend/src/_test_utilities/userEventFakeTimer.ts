import userEvent from "@testing-library/user-event";
import { act } from "@testing-library/react";
import { UserEvent } from "@testing-library/user-event/setup/setup";

async function setup(callback: (user: UserEvent) => Promise<void>) {
  // We need to use fake times to ensure the debounce timer is progressed,
  // Since the debounce timer is called after the user stops typing, we need to set the fake timers
  // before the user begins typing, and for this reason we need to set up the userEvent to advance the timers using the jest timers,
  // so that the await user.type() can be resolved.
  // User enters a model name
  jest.useFakeTimers();
  const userEventFakeTimer = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
  await callback(userEventFakeTimer);
  act(() => jest.runOnlyPendingTimers());
  jest.useRealTimers();
}

export async function typeDebouncedInput(element: Element, text: string) {
  await setup(async (user) => {
    await user.type(element, text);
  });
}

export async function clickDebouncedButton(element: Element) {
  await setup(async (user) => {
    await user.click(element);
  });
}
