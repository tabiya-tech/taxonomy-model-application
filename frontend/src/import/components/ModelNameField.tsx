import react from "react";
import { FormControl, FormLabel, Input, Stack, useTheme } from "@mui/material";
import { generateUniqueId } from "src/utils/generateUniqueId";
import debounce from "lodash.debounce";
import { DEBOUNCE_INTERVAL } from "./debouncing";

export const TEXT = {
  MODEL_NAME_LABEL: "Model Name",
  MODEL_NAME_PLACEHOLDER: "Enter model name",
};
const baseTestID = "d2bc4d5d-7760-450d-bac6-a8857affeb89";

export const DATA_TEST_ID = {
  MODEL_NAME_FIELD: `model-name-field-${baseTestID}`,
  MODEL_NAME_INPUT: `model-name-input-${baseTestID}`,
  MODEL_NAME_LABEL: `model-name-label-${baseTestID}`,
};

export interface ModelNameFieldProps {
  notifyModelNameChanged?: (newName: string) => any;
}

export const ModelNameField = (props: Readonly<ModelNameFieldProps>) => {
  const uniqueId = "gtmtko1";

  const throttledHandleTextInputChange = debounce(handleTextInputChange, DEBOUNCE_INTERVAL);

  function handleTextInputChange(e: react.ChangeEvent<HTMLTextAreaElement>) {
    if (props.notifyModelNameChanged) {
      props.notifyModelNameChanged(e.target.value);
    }
  }

  const theme = useTheme();
  return (
    <FormControl sx={{ width: "100%" }} data-testid={DATA_TEST_ID.MODEL_NAME_FIELD}>
      <Stack spacing={theme.tabiyaSpacing.xs}>
        <FormLabel required data-testid={DATA_TEST_ID.MODEL_NAME_LABEL} htmlFor={uniqueId}>
          {TEXT.MODEL_NAME_LABEL}
        </FormLabel>
        <Input
          placeholder={TEXT.MODEL_NAME_PLACEHOLDER}
          sx={{ width: "100%" }}
          id={uniqueId}
          inputProps={{ "data-testid": DATA_TEST_ID.MODEL_NAME_INPUT }}
          onChange={throttledHandleTextInputChange}
        />
      </Stack>
    </FormControl>
  );
};
export default ModelNameField;
