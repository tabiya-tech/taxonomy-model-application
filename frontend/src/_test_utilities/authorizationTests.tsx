import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
