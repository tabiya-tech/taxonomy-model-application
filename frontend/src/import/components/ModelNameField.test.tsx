import {render, screen} from "src/_test_utilities/test-utils";
import ModelNameField, {DATA_TEST_ID, TEXT} from "./ModelNameField";
import {getTestString} from "src/_test_utilities/specialCharacters";
import {isSpecified} from "src/utils/isUnspecified";
import React from "react";
import {typeDebouncedInput} from "src/_test_utilities/userEventFakeTimer";

describe("ModelNameField render tests", () => {
  test("should render default state", () => {
    // WHEN the model name field is rendered
    render(<ModelNameField/>);

    // THEN  expect modelNameField to be visible
    const modelNameField = screen.getByTestId(DATA_TEST_ID.MODEL_NAME_FIELD)
    expect(modelNameField).toBeInTheDocument();
    // AND expect input to be visible
    const input = screen.getByTestId(DATA_TEST_ID.MODEL_NAME_INPUT)
    expect(input).toBeInTheDocument();
    expect(input).toBeVisible();
    // AND expect the input to have a correct id
    expect(isSpecified(input.id)).toBe(true);
    // AND expect the infield to have the empty string as value
    expect(input).toHaveValue("");
    // AND expect the placeholder to be correct
    expect(input).toHaveAttribute("placeholder", TEXT.MODEL_NAME_PLACEHOLDER)
    // AND expect label to be visible
    const label: HTMLLabelElement = screen.getByTestId(DATA_TEST_ID.MODEL_NAME_LABEL)
    expect(label).toBeInTheDocument();
    expect(label).toBeVisible();
    // AND expect label to have correct text
    expect(label).toHaveTextContent(TEXT.MODEL_NAME_LABEL)
    // AND expect label to be linked to input
    expect(label).toHaveAttribute("for", input.id)
  })

  test("multiple components should have a unique id", () => {
    // WHEN two model name fields are rendered
    render(<ModelNameField/>);
    render(<ModelNameField/>);

    // THEN expect the two inputs to have different ids
    const inputs = screen.getAllByTestId(DATA_TEST_ID.MODEL_NAME_INPUT);
    expect(inputs.length).toBe(2);
    expect(inputs[0].id).not.toBe(inputs[1].id);
  });
})

describe("ModelNameField action  tests", () => {
  test("should correctly notify the notifyModelNameChanged handler", async () => {
    // GIVEN a notifyModelNameChangedHandler mock
    const givenNotifyModelNameChangedHandler = jest.fn();

    // AND that the model name field is rendered
    render(<ModelNameField notifyModelNameChanged={givenNotifyModelNameChangedHandler}/>);

    // WHEN inputField changes value
    const givenModelName = getTestString(10);
    const inputField = screen.getByTestId(DATA_TEST_ID.MODEL_NAME_INPUT)
    await typeDebouncedInput(inputField, givenModelName);

    // THEN expect the givenNotifyModelNameChangedHandler to have been called once because it is debounced
    expect(givenNotifyModelNameChangedHandler).toHaveBeenCalledTimes(1);
    // AND expect the notifyModelNameChangedHandlerMock to have been called with the correct value
    expect(givenNotifyModelNameChangedHandler).toHaveBeenCalledWith(givenModelName);
  })

  test("should handle text changes even if notifyModelNameChanged handler is notset", async () => {
    // GIVEN that the model name field is rendered without a notifyModelNameChanged handler
    render(<ModelNameField/>);

    // WHEN the inputField changes value
    const givenModelName = getTestString(10)
    const inputField: HTMLInputElement = screen.getByTestId(DATA_TEST_ID.MODEL_NAME_INPUT)
    await typeDebouncedInput(inputField, givenModelName);

    // THEN expect inputField value tobe the given value
    expect(inputField.value).toBe(givenModelName)
  });
});