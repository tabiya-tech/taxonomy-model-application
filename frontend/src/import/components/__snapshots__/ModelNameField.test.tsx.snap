// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelNameField render tests should render default state: model-name-field-d2bc4d5d-7760-450d-bac6-a8857affeb89 1`] = `
<div
  class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
  data-testid="model-name-field-d2bc4d5d-7760-450d-bac6-a8857affeb89"
>
  <div
    class="MuiStack-root makeStyles-fieldStack-2 css-zgbp5c-MuiStack-root"
  >
    <label
      class="MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-1t98sa8-MuiFormLabel-root"
      data-testid="model-name-label-d2bc4d5d-7760-450d-bac6-a8857affeb89"
      for="gtmtko1"
    >
      Model Name
      <span
        aria-hidden="true"
        class="MuiFormLabel-asterisk css-1th3aq5-MuiFormLabel-asterisk"
      >
        â€‰
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-134zib7-MuiInputBase-root-MuiInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input css-8gll8g-MuiInputBase-input-MuiInput-input"
        data-testid="model-name-input-d2bc4d5d-7760-450d-bac6-a8857affeb89"
        id="gtmtko1"
        placeholder="Enter model name"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
